java/src/Ice/src/main/java/com/zeroc/IceInternal/Instance.java

```diff
@@ -720,90 +720,82 @@ public Class<?> apply(String typeId)
         Class<?> c = null;
 
         //
-        // To convert a Slice type id into a Java class, we do the following:
+        // To convert a Slice type ID into a Java class, the following steps are taken:
         //
-        // 1. Convert the Slice type id into a classname (e.g., ::M::X -> M.X).
-        // 2. If that fails, extract the top-level module (if any) from the type id
-        //    and check for a Package property. If found, prepend the property
-        //    value to the classname.
-        // 3. If that fails, check for an Default.Package property. If found,
-        //    prepend the property value to the classname.
+        // 1. Convert the Slice type ID into a classname (e.g., ::M::X becomes M.X).
+        // 2. Check if the application has defined any package prefixes for the top-level module (e.gm "M").
+        //    Attempt to resolve the <package-prefix>+<class-name> in order trying each defined package
+        //    prefix.
+        // 3. If the above step fails, search for a Default.Package property. If found,
+        //    prepend its value to the classname. Otherwise, attempt to use the
+        //    classname directly.
         //
         String fullyQualifiedClassName;
-        boolean addClass = false;
 
         //
         // See if we've already translated this type ID before.
         //
-        synchronized(this)
+        synchronized (this)
         {
             fullyQualifiedClassName = _sliceTypeIdToClassMap.get(typeId);
         }
 
         //
-        // It's a new type ID, so first convert it into a Java class name.
+        // The fully qualified class name was already resolved, keep using it.
         //
-        if(fullyQualifiedClassName == null)
+        if (fullyQualifiedClassName != null)
         {
-            fullyQualifiedClassName = com.zeroc.Ice.Util.typeIdToClass(typeId);
-            addClass = true;
```

Comment:
`addClass` is no longer required as we return earlier if this is a class we have previously resolved.